vowpalwabbit/OjaNewton.cc

```diff
@@ -52,51 +52,53 @@ struct OjaNewton {
     bool normalize;
     bool random_init;
 
-    void initialize_Z()
-    {
-        size_t stride_shift = all->reg.stride_shift;
-        weight* weights = all->reg.weight_vector;
-
-        uint32_t length = 1 << all->num_bits;
-        if(normalize) { // initialize normalization part
-	    for (uint32_t i = 0; i < length; i++)
-	        weights[(i << stride_shift) + NORM2] = 0.1;
-        }
-
+	void initialize_Z()
+	{  weight_vector& weights = *(all->wv);
+       uint32_t length = 1 << all->num_bits;
+	   if (normalize) { // initialize normalization part
+		 for (weight_vector::iterator i = weights.begin(NORM2); i != weights.end(NORM2); ++i)
+		   *i = 0.1;
+	    }
         if(!random_init) {
             // simple initialization
-            for (int i = 1; i <= m; i++)
-                weights[(i << stride_shift) + i] = 1;
+		  weight_vector::iterator iter = weights.begin(0) + 1;
+		  for (int i = 1; i <= m; ++i, ++iter)
+		  { weight_vector::iterator::w_iter j = iter.begin() + i;
+			*j = 1;
+		  }
         }
-	else {
+	    else {
             // more complicated initialization: orthgonal basis of a random matrix
 	    const double PI = 3.1415927;
-
-            for (uint32_t i = 0; i < length; i++) {
-                for (int j = 1; j <= m; j++) {
-	           double r1 = frand48();
-	           double r2 = frand48();
-	           weights[(i << stride_shift) + j] = sqrt(-2 * log(r1)) * cos(2 * PI * r2);
-	        }
-	    }
-
+		for (weight_vector::iterator i = weights.begin(0); i != weights.end(); ++i)
+		  for (weight_vector::iterator::w_iter j = i.begin() + 1; j != i.end(m + 1); ++j)
+		  {  double r1 = frand48();
+		     double r2 = frand48();
+			 *j = sqrt(-2 * log(r1)) * cos(2 * PI * r2);
```

Comment:
This is box-meuller (i.e. gaussian random variable) initialization.  Making this work in a reproducible sparse fashion is a little bit tricky but possible.  You can seed the PRG for an entry with the entry's index.